okhttp-dnsoverhttps/src/main/java/okhttp3/dnsoverhttps/DnsOverHttps.java

```diff
@@ -131,37 +133,121 @@ public boolean resolvePublicAddresses() {
   }
 
   private List<InetAddress> lookupHttps(String hostname) throws UnknownHostException {
+    List<Call> networkRequests = new ArrayList<>(2);
+    List<Response> responses = new ArrayList<>(2);
+    List<Exception> failures = new ArrayList<>(2);
+
+    // TODO Optimise for synchronous unmerged case when IPv4 only
+
+    buildRequest(hostname, networkRequests, responses, DnsRecordCodec.TYPE_A);
+
+    if (includeIPv6) {
+      buildRequest(hostname, networkRequests, responses, DnsRecordCodec.TYPE_AAAA);
+    }
+
+    executeRequests(networkRequests, responses, failures);
+
+    return readResponses(hostname, responses, failures);
+  }
+
+  private void buildRequest(String hostname, List<Call> networkRequests, List<Response> responses,
+      int type) {
+    Request ipv4Request = buildRequest(hostname, type);
+    Response ipv4Response = getCacheOnlyResponse(ipv4Request);
+
+    if (ipv4Response != null) {
+      responses.add(ipv4Response);
+    } else {
+      networkRequests.add(client.newCall(ipv4Request));
+    }
+  }
+
+  private void executeRequests(List<Call> networkRequests, final List<Response> responses,
+      final List<Exception> failures) {
+    final CountDownLatch latch = new CountDownLatch(networkRequests.size());
+
+    for (Call call : networkRequests) {
+      call.enqueue(new Callback() {
+        @Override public void onFailure(Call call, IOException e) {
+          synchronized (failures) {
+            failures.add(e);
+          }
+          latch.countDown();
+        }
+
+        @Override public void onResponse(Call call, Response response) throws IOException {
+          synchronized (responses) {
+            responses.add(response);
```

Comment:
Worth decoding the response body here? Otherwise we're doing blocking reads in serial not in parallel.